File: Java/src/com/leetcode/Medium/UniqueBinarySearchTreesII.java
package com.leetcode.Medium;

import java.util.*;

// 95. Unique Binary Search Trees II
public class UniqueBinarySearchTreesII {
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) {
            this.val = val;
        }
    }

    public List<TreeNode> generateTrees(int n) {
        if (n == 0)
            return new ArrayList<>();

        return helper(1, n);
    }

    private List<TreeNode> helper(int m, int n) {
        List<TreeNode> result = new ArrayList<>();
        if (m > n) {
            result.add(null);
            return result;
        }

        for (int i = m; i <= n; i++) {
            List<TreeNode> ls = helper(m, i - 1);
            List<TreeNode> rs = helper(i + 1, n);
            for (TreeNode l : ls) {
                for (TreeNode r : rs) {
                    TreeNode node = new TreeNode(i);
                    node.left = l;
                    node.right = r;
                    result.add(node);
                }
            }
        }

        return result;
    }
}

File: JavaScript/Medium/UniqueBinarySearchTreesII.js
/**
 * 95. Unique Binary Search Trees II
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
const generateTrees = function(n) {
    if (n === 0)
        return [];

    return helper(1, n);
};

const helper = function (m, n) {
    let result = [];
    if (m > n) {
        result.push(null);
        return result;
    }
    
    for (let i = m; i <= n; i++) {
        let ls = helper(m, i - 1);
        let rs = helper(i + 1, n);
        for (let l of ls) {
            for (let r of rs) {
                let node = new TreeNode(i);
                node.left = l;
                node.right = r;
                result.push(node);
            }
        }
    }
    
    return result;
};